### PR TITLE
GraphQL: Dynamic block support

### PR DESCRIPTION
**Pull request by keharper**
## Purpose of this pull request

This pull request (PR) adds the `dynamicBlocks` query.

Related code PR: https://github.com/magento/partners-magento2ee/pull/421

## Affected DevDocs pages

-  ...

## Links to Magento source code

-  ...

whatsnew
Added the `dynamicBlocks` query, which returns the contents of dynamic blocks that match the specified filters.

### CODE REVIEW

File: src/guides/v2.4/graphql/queries/dynamic-blocks.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+---
+group: graphql
+title: dynamicBlocks query
+---
+
+Dynamic blocks can be inserted in CMS pages and blocks using widgets. The `dynamicBlocks` query returns the content of dynamic blocks that implement the Dynamic Blocks Rotator widget.
```
Comment (cpartica) [inline src/guides/v2.4/graphql/queries/dynamic-blocks.md:6]: we have to make it more obvious that Dynamic Blocks have to be defined in the Dynamic Blocks section of the admin and Dynamic Blocks Rotator is a widget inserted in a CMS page.
Purpose of the new gql query that we added is to grab content of what the widget is returning by parsing some html and extract those parameters that you describe further

That means doing some reference of how the Dynamic Blocks Rotator widget looks like in html when is returned by the cmsPage query.


You can also define widgets

**Review by cpartica — CHANGES_REQUESTED** [review]

File: src/guides/v2.4/graphql/queries/dynamic-blocks.md (hunk 1 above)
Comment (davemacaulay) [reply src/guides/v2.4/graphql/queries/dynamic-blocks.md:6]: I would promote linking to the user guides for how to setup Dynamic Blocks in this section to help provide context. Also note we renamed "Banner" to "Dynamic Block" in 2.3.1 or 2.3.0, so may need to briefly mention that too.

**Review by davemacaulay — CHANGES_REQUESTED** [review]

File: src/guides/v2.4/graphql/queries/dynamic-blocks.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+dynamic_blocks(
+    input: DynamicBlocksFilterInput
+    pageSize: Int
+    currentPage: Int
+    ): DynamicBlocks!
+```
+
+## Example usage
+
+The following query returns all dynamic blocks of type `SPECIFIED`. The returned item is a a dynamic block containing only text. The second item contains a PNG file.
```
Comment (dobooth) [inline src/guides/v2.4/graphql/queries/dynamic-blocks.md:34]:
```suggestion
The following query returns all dynamic blocks of type `SPECIFIED`. The returned item is a dynamic block containing only text. The second item contains a PNG file.
```

File: src/guides/v2.4/graphql/queries/dynamic-blocks.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      },
+      "total_count": 2
+    }
+  }
+}
+```
+
+**cmsPage query response:**
+
+The [`cmsPage` query]({{page.baseurl}}/graphql/queries/cms-page.html) would return the following code for a CMS page that contains the dynamic block in this example with the `uid` of `MQ==`. The response has been reformatted for readability.
```
Comment (dobooth) [inline src/guides/v2.4/graphql/queries/dynamic-blocks.md:94]:
```suggestion
The [`cmsPage` query]({{page.baseurl}}/graphql/queries/cms-page.html) returns the following code for a CMS page that contains the dynamic block; in this example, a block with the `uid` of `MQ==`. The response has been reformatted for readability.
```

File: src/guides/v2.4/graphql/queries/dynamic-blocks.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+```
+
+## Input attributes
+
+Attribute | Data type | Description
+--- | --- | ---
+`dynamic_block_uids` | [ID] | An array of dynamic block UIDs to filter on
+`locations` | [DynamicBlockLocationEnum] | An array indicating the locations the dynamic block can be placed. The possible values are CONTENT, HEADER, FOOTER, LEFT, and RIGHT. If this attribute is not specified, the query returns all locations
+`type` | DynamicBlockTypeEnum! | A value indicating the type of dynamic block to filter on. The possible values are CART_PRICE_RULE_RELATED, CATALOG_PRICE_RULE_RELATED, and SPECIFIED
+## Output attributes
```
Comment (dobooth) [inline src/guides/v2.4/graphql/queries/dynamic-blocks.md:131]: Space between table and header.

File: src/guides/v2.4/graphql/queries/dynamic-blocks.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+The `DynamicBlocks` output object contains the following attributes.
```
Comment (dobooth) [inline src/guides/v2.4/graphql/queries/dynamic-blocks.md:133]:
```suggestion
The `DynamicBlocks` output object contains the following attributes:
```
Same with the following two.

**Review by dobooth — CHANGES_REQUESTED** [review]

**Review by dobooth — APPROVED** [review]

File: src/guides/v2.4/graphql/queries/dynamic-blocks.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+group: graphql
+title: dynamicBlocks query
+---
+
+The `dynamicBlocks` query returns a list of dynamic blocks that have been placed in a Dynamic Blocks Rotator inline widget and meet the specified criteria.
+
+{:.bs-callout-info}
+The Banner functionality was removed from Magento 2.4.0 and replaced with dynamic blocks.
+
+When a [Dynamic Blocks Rotator inline widget is created](https://docs.magento.com/user-guide/cms/dynamic-blocks.html), the Magento administrator can select the following options:
```
Comment (cpartica) [inline src/guides/v2.4/graphql/queries/dynamic-blocks.md:11]: link to https://docs.magento.com/user-guide/cms/dynamic-blocks-rotate.html instead

File: src/guides/v2.4/graphql/queries/dynamic-blocks.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+*  **Specified Dynamic Blocks**
+*  **Cart Price Rule Related**
+*  **Catalog Price Rule Related**
+
+Widgets defined with the **Specified Dynamic Blocks** option affect CMS page rendering. The other two options are used for rendering cart, product, and catalog pages and are not applicable for PWA applications. Therefore, in most cases, your query should assign the value of `SPECIFIED` to the `type` input attribute.
```
Comment (cpartica) [inline src/guides/v2.4/graphql/queries/dynamic-blocks.md:17]: in most cases for a CMS page

you also should add that Cart Price Rule Related
Catalog Price Rule Related don't really have direct use cases for CMS page

**Review by cpartica — CHANGES_REQUESTED** [review]

**Comment by keharper** [comment]
running tests

**Review by cpartica — APPROVED** [review]

**Comment by keharper** [comment]
running tests

**Pull request merged**